src/client/Vad.ts

```diff
@@ -0,0 +1,188 @@
+
+/**
+ *  average frequency-based VAD
+ *  logic from https://github.com/Jam3/voice-activity-detection
+ */
+export default class Vad {
+  analyser: AnalyserNode
+  private config: VadConfig
+
+  private baseLevel: number
+  private freqBins: Uint8Array
+  private voiceScale: number
+  private activityCounter: number
+
+  private riseFrames: number
+  private fallFrames: number
+  private maxActiveFrames: number
+
+  private envFreqRange: number[]
+  private prevVadState: boolean
+  private curVadState: boolean
+  private vadActive: boolean
+  private captureTimeout: number
+
+  constructor(audioContext: AudioContext, config: SpeechConfig) {
+    this.config = config.vad || DefaultVadConfig
+
+    this.baseLevel = 0.3
+    this.voiceScale = 1
+
+    this.activityCounter = 0
+    this.riseFrames = 0
+    this.fallFrames = 0
+    this.maxActiveFrames = this.config.maxActive
+
+    this.envFreqRange = []
+    const fftWidth = config.fftWidth || 512
+    this.freqBins = new Uint8Array(fftWidth)
+    this.prevVadState = false
+    this.curVadState = false
+    this.vadActive = false
+    this.captureTimeout = 1000
+
+    this.analyser = audioContext.createAnalyser()
+    this.analyser.smoothingTimeConstant = this.config.smoothingTimeConstant
+    this.analyser.fftSize = fftWidth
+  }
+
+  /**
+   * connects an AnalyserNode between the specified source and script processor
+   * in the audio context so that incoming data can be analyzed for voice
+   * activity. The node must be connected before `vadStatus` will work.
+   */
+  connect(source: MediaStreamAudioSourceNode, processor: ScriptProcessorNode) {
+    source.connect(this.analyser)
+    this.analyser.connect(processor)
+
+    // we'll receive `bufferSize` samples at a time, so use this to determine
+    // activity parameters
+    const frameSize = processor.bufferSize
+    const msPerFrame = frameSize / (source.context.sampleRate / 1000)
+    this.riseFrames = Math.floor(this.config.riseDelay / msPerFrame)
+    this.fallFrames = Math.floor(this.config.fallDelay / msPerFrame)
+    this.maxActiveFrames = Math.floor(this.config.maxActive / msPerFrame)
+  }
+
+  disconnect() {
+    this.analyser.disconnect()
+  }
+
+  /**
+   * Analyzes sound for the amount of time specified by
+   * `noiseCaptureDuration` to determine a base level of
+   * background noise to discount during voice activity detection.
+   *
+   * The VAD can be run without calibration, but different browsers
+   * and input methods can report different sample values for the same
+   * level of background noise, so calibration before first use is recommended.
+   *
+   * The VAD must be `connect`ed before it can be calibrated
+   */
+  calibrate(onComplete: () => void) {
+    if (this.analyser.numberOfOutputs === 0) {
+      throw new Error('VAD is not connected')
+    }
+
+    setTimeout(() => {
+      this.setLevels()
+      onComplete()
+    }, this.config.noiseCaptureDuration)
```

Comment:
This value was originally measured in seconds, so we might want to change the default in `types.ts` to reflect that and capture more audio for calibration...that said, I tried it, and in a quiet room I didn't have a problem. Might be different with background noise.

Reply:
oh, no I just forgot the conversion.